analysis/3_coregister-and-source-recon-bulk.py

```diff
@@ -0,0 +1,240 @@
+"""Coregistration and source reconstruction 
+Created AXL 28/10/24
+
+The script should be first run for all subjects. Then for subjects
+whose coregistration looked a bit off we re-run this script just for that
+particular subject.
+
+This does not use OSL batch utils/config-file setup.
+"""
+
+############## ------- Import packages and define functions needed ---------- ############
+
+import os
+import os.path as op
+import osl
+from pprint import pprint
+from osl import utils
+from osl import source_recon
+import numpy as np
+import coinsmeg_data as coinsmeg
+from IPython.display import HTML, display
+import mne
+from glob import glob
+import pathlib
+
+# functions
+def copy_polhemus_files(recon_dir, subject, preproc_file, smri_file, logger):
+    polhemus_headshape = np.loadtxt(op.join(polhemus_dir, 'polhemus_headshape.txt'))
+    polhemus_nasion = np.loadtxt(op.join(polhemus_dir, 'polhemus_nasion.txt'))
+    polhemus_rpa = np.loadtxt(op.join(polhemus_dir, 'polhemus_rpa.txt'))
+    polhemus_lpa = np.loadtxt(op.join(polhemus_dir, 'polhemus_lpa.txt'))
+    
+    #  Get coreg filenames
+    filenames = source_recon.rhino.get_coreg_filenames(recon_dir, subject)
+
+    # Save
+    np.savetxt(filenames["polhemus_nasion_file"], polhemus_nasion)
+    np.savetxt(filenames["polhemus_rpa_file"], polhemus_rpa)
+    np.savetxt(filenames["polhemus_lpa_file"], polhemus_lpa)
+    np.savetxt(filenames["polhemus_headshape_file"], polhemus_headshape)
+
+############## ------- Directories ---------- ############
+
+# Directories
+DATA_DIR = coinsmeg.RAW_DIR # this is the same as BASE_DIR as raw data is stored in the base directory
+preproc_dir = coinsmeg.PREPROCESSED_DIR
+recon_dir = op.join(coinsmeg.DERIVATIVES_DIR, "recon")
+
+# Get subjects
+subs = []
+
+data_sub_folders = sorted(filter(lambda path: "sub-" in path, glob(DATA_DIR + '/*'))) # returns a list of paths e.g., '/ohba/pi/lhunt/datasets/coins-meg_data/sub-22'
+
+for subject in data_sub_folders:
+    subs.append(pathlib.Path(subject).stem) # subs is now a list of ['sub-01', 'sub-02', ...]
+
+# Create a list with '_run-X' appended for each subject
+sub_run_combos = []
+for sub in subs:
+    for run in range(1, 5):  # Loop from 1 to 4
+        sub_run_combos.append(f"{sub}_run-{run}") # sub_run_combos is now a list of ['sub-01_run-1', 'sub-01_run-2', ...]
+
+print(sub_run_combos)
+
+# test with only two sub/run combos
+sub_run_combos =['sub-04_run-1','sub-04_run-2']
```

Comment:
This can now be removed from the script - we want it to run for all subjects/runs